lxd/containers.go

```diff
@@ -51,90 +54,227 @@ func containersPost(d *Daemon, r *http.Request) Response {
+	return &asyncResponse{run: run, containers: []string{req.Name}}
+}
+
+func removeContainerPath(name string) {
+	cpath := shared.VarPath("lxc", name)
+	err := os.RemoveAll(cpath)
+	if err != nil {
+		shared.Debugf("Error cleaning up %s: %s\n", cpath, err)
+	}
+}
+
+func extractShiftRootfs(uuid string, name string, d *Daemon) error {
+	cleanup := func(err error) error {
+		removeContainerPath(name)
+		dbRemoveContainer(name)
+		return err
 	}
 
 	/*
-	 * Actually create the container
+	 * We want to use archive/tar for this, but that doesn't appear
+	 * to be working for us (see lxd/images.go)
+	 * So for now, we extract the rootfs.tar.xz from the image
+	 * tarball to /var/lib/lxd/lxc/container/rootfs.tar.xz, then
+	 * extract that under /var/lib/lxd/lxc/container/rootfs/
 	 */
-	run := shared.OperationWrap(func() error { return c.Create(opts) })
-	return &asyncResponse{run: run, containers: []string{req.Name}}
+	dpath := shared.VarPath("lxc", name)
+	fmt.Printf("uuid is %s\n", uuid)
+	imagefile := shared.VarPath("images", uuid)
+	output, err := exec.Command("tar", "-C", dpath, "-Jxf", imagefile, "rootfs.tar.xz").Output()
+	if err != nil {
+		fmt.Printf("Untar of image: Output %s\nError %s\n", output, err)
+		return cleanup(err)
+	}
+
+	rpath := shared.VarPath("lxc", name, "rootfs")
+	tarfile := shared.VarPath("lxc", name, "rootfs.tar.xz")
+	output, err = exec.Command("tar", "-C", rpath, "--numeric-owner", "-Jxpf", tarfile).Output()
+	if err != nil {
+		fmt.Printf("Untar of rootfs: Output %s\nError %s\n", output, err)
+		return cleanup(err)
+	}
+
+	err = d.id_map.ShiftRootfs(rpath)
+	if err != nil {
+		fmt.Printf("Shift of rootfs %s failed: %s\n", rpath, err)
+		return cleanup(err)
+	}
+
+	/* Set an acl so the container root can descend the container dir */
+	acl := fmt.Sprintf("%d:rx", d.id_map.Uidmin)
+	_, err = exec.Command("setfacl", "-m", acl, dpath).Output()
+	if err != nil {
+		fmt.Printf("Error adding acl for container root: start will likely fail\n")
+	}
+
+	return nil
+}
+
+func dbRemoveContainer(name string) {
+	certdbname := shared.VarPath("lxd.db")
+	db, err := sql.Open("sqlite3", certdbname)
+	if err != nil {
+		return
+	}
+	defer db.Close()
+
+	_, _ = db.Exec("DELETE FROM containers WHERE name=?", name)
+}
+
+func dbGetImageId(image string) (int, string, error) {
+	certdbname := shared.VarPath("lxd.db")
+	db, err := sql.Open("sqlite3", certdbname)
+	if err != nil {
+		return 0, "", err
+	}
+	defer db.Close()
+
+	/* todo - look at aliases */
+	rows, err := db.Query("SELECT id, fingerprint FROM images WHERE fingerprint=?", image)
+	if err != nil {
+		return 0, "", err
+	}
+	defer rows.Close()
+	id := -1
+	var uuid string
+	for rows.Next() {
+		rows.Scan(&id, &uuid)
+	}
+	if id == -1 {
+		return 0, "", fmt.Errorf("Unknown image")
+	}
+	return id, uuid, nil
+}
+
+func dbGetContainerId(db *sql.DB, name string) (int, error) {
+	rows, err := db.Query("SELECT id FROM containers WHERE name = ?", name)
```

Comment:
The extra spaces around the equal sign doesn't match the rest of the code (and is considered invalid by some DB engines).